Registry mvp docs (#689)
wandb/docs: docs/guides/registry/create_registry.md

---
displayed_sidebar: default
---

# Create a custom registry

> Create a custom registry within an organization. Custom registries are particularly useful for organizing project-specific requirements that differ from the default, core registries that W&B provides. For example, you might need a registry dedicated to storing experiment configurations or hyperparameters for a specific research project. 

Reviewer: i would use different examples here, such as evaluation datasets, fine-tuning datasets, code, jobs, etc.

Author: This is a dumb question: will it not be a common use case for  users to use the default Dataset registry for (e.g.) evaluation datasets, fine-tuning datasets? In this case, they would create a collection for this task?

Author: Which brings me to: What is the canonical ml workflow example that will resonate with new users and custom registries?

Reviewer: Not a dumb question - yes, technically, Evaluation Datasets or Benchmark datasets VS. training datasets - they're all datasets. But they are used in different contexts and they're versioned independently of each other. 

Training Dataset:
- Model reproducibility.
- Track data changes for continuous improvement.
- Facilitate experimentation by comparing models on different data versions.
- Format: Large, diverse text data in formats like plain text, JSON, CSV. May include raw text, annotations.

Benchmark Dataset:
- Consistent evaluation over time.
- Fair comparison between models or versions.
- Track improvements in evaluation quality without affecting previous benchmarks.
- Format: Smaller, focused, high-quality annotated data in formats like JSON, CSV.

**What would different collections be called within each registry?**

Training Dataset (in Dataset Registry) (might contain raw data that is scraped): 
- Raw_User_Browsing_History
- Raw_Product_Descriptions_Reviews
- Raw_Transaction_Logs

Benchmark Dataset (curated for validation to produce a score):
- Curated_User_Profiles_Known_Preferences
- Curated_Product_Lists_Expert_Recommendations
- Curated_Historical_Recommendation_Outcomes

And then for fine-tuning, you might have a registry that has these tasks:

- Fine_Tuning_Code_Completion_Dataset
- Fine_Tuning_Code_Documentation_Dataset
- Fine_Tuning_Bug_Reports_Resolution_Dataset
- Fine_Tuning_Constitutional_Law_Cases_Dataset
- Fine_Tuning_Legal_Documents_Summarization_Dataset
- Fine_Tuning_Patent_Law_Analysis_Dataset
- Fine_Tuning_Healthcare_EHR_Data

Within those tasks, there could still be versioning that happens (you delete corrupted data, add a new corpus of data for that task)

Reviewer: **Project: Developing an Ensemble of LLMs to Answer Questions on Medical Insurance Policies**

1. **Registry Name: Raw_Datasets**
   - **Purpose**: Store the initial raw data collected from various sources before any preprocessing or cleaning.
   - **Contents**:
     - `Raw_User_Queries_Data`: Unprocessed text data gathered from user interactions, customer service logs, or other sources where users ask questions about medical insurance policies.
     - `Raw_Insurance_Policies`: Collected medical insurance policy documents, including terms and conditions, coverage details, and exclusions.

2. **Registry Name: Training_Datasets**
   - **Purpose**: Contain the data that has been cleaned, preprocessed, and is ready for training the model.
   - **Contents**:
     - `User_Query_Insurance_Answer_Training_Data`: Pairs of user questions and corresponding answers extracted from insurance policies. This dataset is used to teach the LLM how to accurately respond to queries about medical insurance policies.

3. **Registry Name: Benchmark_Datasets**
   - **Purpose**: Hold curated datasets used to evaluate the performance of the trained models. These datasets are designed to provide a consistent measure of model accuracy and effectiveness.
   - **Contents**:
     - `User_Query_Insurance_Answer_Test_Data`: A set of user questions and corresponding expert-validated answers that the model has never seen during training. This dataset is used to test the model’s ability to accurately answer new questions about medical insurance policies.

4. **Registry Name: Fine_Tuned_Models**
   - **Purpose**: Store the versions of models that have been fine-tuned on the specific training datasets for different parts of the ensemble model pipeline.
   - **Contents**:
     - `Fine_Tuned_Query_Classifier_v1`: First version of the model fine-tuned to classify the type of user query.
     - `Fine_Tuned_Policy_Extractor_v1`: First version of the model fine-tuned to extract relevant sections from insurance policies.
     - `Fine_Tuned_Response_Generator_v1`: First version of the model fine-tuned to generate user-friendly responses based on extracted policy sections.

5. **Registry Name: Deployment_Configs**
   - **Purpose**: Store configurations and parameters necessary for deploying the trained models into a production environment.
   - **Contents**:
     - `Query_Classifier_Deployment_Config`: Configuration files for deploying the query classification model.
     - `Policy_Extractor_Deployment_Config`: Configuration files for deploying the policy extraction model.
     - `Response_Generator_Deployment_Config`: Configuration files for deploying the response generation model.

Each registry serves a specific purpose in the machine learning pipeline, from initial data collection to final model deployment.

Author: Ahh. OK, so the mapping is to have registry for ea. step of the ML pipeline. 

The mental model I had was a one stop shop for all data (in the case of registry with dataset type) related to a specific ML task. For e.g. Suppose I want to train a CNN to classify/identify earthquakes from seismogram recordings. In this case:

* Registry name: Realtime earthquake data (seismograms)
   * Collections:
      * Northern California earthquake data 
         * V0 raw dataset
         * V1 processed dataset
      * Southern California earthquake data 
         * V0 raw dataset
         * V1 processed dataset
      * Training data
         * Combined, processed NorCal and SoCal dataset: v0  
      * Test data
         *  Subset of combined, processed NorCal and SoCal dataset: v0  

So, in the example you have, you have a separate registry for raw and processed datasets. (Whereas I just bunch it into one).

For the model part:

* Registry name: CNN Earthquake Classifier
   * Collections:
      * CNN (2 layers only)
         * model:v0 
         * model:v1
      * CNN (4 layers)
         * model:v0 
         * model:v1
      * Transfer learning CNN using MNIST
         * model:v0 
         * model:v1

Author: Anyhow, I used one of the examples you listed above.

:::caution

> W&B Registries is in active development. Currently, you can not remove an artifact type once it is defined.  

Reviewer: I'm confused how these 2 sentences are related (this is unrelated to registries being in active development). 

```suggestion
W&B Registries is in active development. Currently, you can not remove an artifact type once it is added and the registry settings are saved.  
```
^ why is the note on artifact types here? this should be noted in context when we talk about editing types

Author: It's referenced in step #6:

"6. Select either **All types** or **Specify types** from the **Accepted artifacts type** dropdown."

:::

1. Navigate to the Registries App in the W&B App UI.
2. Within **Custom registries**, click on the **Create registry** button.
3. Provide a name for your registry in the **Name** field.
4. Optionally provide a description about the registry.

> 5. Select who can view the registry from the **Registry visibility** dropdown. See [LINK] for more information on how to limit visibility.

Reviewer: ```suggestion
5. Select who can view the registry from the **Registry visibility** dropdown. See [LINK] for more information on what the available options are for registry visibility.
```

> 6. Select either **All types** or **Specify types** from the **Accepted artifacts type** dropdown.

Reviewer: it's not going to be obvious to new users who don't know what artifacts are what accepted artifact types are. I think we should have a section that talks about "Defining a registry's accepted types" - this should explain what types are, how they are set when creating artifacts, what are the implications of selecting All types vs specify types, why you should standardize accepted types (i.e so a user doesn't link a model artifact to dataset registry). Types is an important mechanism for how a user organizes their registry and enforces that the right things go in the right registries.

7. (If you select **Specify types**) Add one or more artifact types that your registry permits.
8. Click on the **Create registry** button. 

![](/images/registry/create_custom_registry.png)

> The preceding image shows a custom registry called "Forecast" that a user is about to create. The registry is set to **Restricted** [LINK] which means that only members that are manually added to the "Forecast" registry will have access to this registry. In addition, only  `hf-model` artifact types can be added to the "Forecast" registry (see the **Specify types** field). 

Reviewer: can pick a registry name and type example that makes a bit more sense - i.e Benchmark Datasets, and type is "hf-dataset"

Author: I'm limited to the prototype example (it's the only one we have w/ this workflow) 😬

Author: We can update once we have another example

Reviewer: ahhhh yes ok gotcha gotcha - commented in [this](https://weightsandbiases.slack.com/archives/C01DX8LRZEJ/p1718939746641269?thread_ts=1718318901.046779&cid=C01DX8LRZEJ) thread so you can start using feature

## Restrict visibility to a registry
<!-- Who can do this? -->
Restrict who can view and access a custom registry. You can restrict visibility to a registry when you create a custom registry or after you create a custom registry. A custom registry can have either restricted or organization visibility. For more information on registry visibilities, see [LINK].

<!-- | Visibility | Description |
| --- | --- |
| Organization | Anyone in the organization can view the registry. |
| Restricted   | Only invited organization members can view and edit the registry.|  -->

> The following steps describe how to restrict the visibility of a custom registry that already exists:

Reviewer: can we just show this in a gif instead of listing out all the steps?

Author: We will use both. Text is required for accessibility reasons.
